cpp/include/utilities/comm_utils.cuh

```diff
@@ -227,6 +227,326 @@ struct device_irecv_tuple_iterator_element_impl<InputIterator, OutputIterator, I
+    auto tuple_element_output_first = thrust::get<I>(output_first.get_iterator_tuple());
+    device_sendrecv_impl<decltype(tuple_element_input_first), decltype(tuple_element_output_first)>(
+      comm,
+      tuple_element_input_first,
+      tx_count,
+      dst,
+      tuple_element_output_first,
+      rx_count,
+      src,
+      stream);
+    device_sendrecv_tuple_iterator_element_impl<InputIterator, OutputIterator, I + 1, N>().run(
+      comm, input_first, tx_count, dst, output_first, rx_count, src, stream);
+  }
+};
+
+template <typename InputIterator, typename OutputIterator, size_t I>
+struct device_sendrecv_tuple_iterator_element_impl<InputIterator, OutputIterator, I, I> {
+  void run(raft::comms::comms_t const& comm,
+           InputIterator input_first,
+           size_t count,
+           int dst,
+           int base_tag,
+           raft::comms::request_t* requests) const
+  {
+  }
+};
+
+template <typename InputIterator, typename OutputIterator>
+std::enable_if_t<thrust::detail::is_discard_iterator<OutputIterator>::value, void>
+device_multicast_sendrecv_impl(raft::comms::comms_t const& comm,
+                               InputIterator input_first,
+                               std::vector<size_t> const& tx_counts,
+                               std::vector<size_t> const& tx_offsets,
+                               std::vector<int> const& tx_dst_ranks,
+                               OutputIterator output_first,
+                               std::vector<size_t> const& rx_counts,
+                               std::vector<size_t> const& rx_offsets,
+                               std::vector<int> const& rx_src_ranks,
+                               cudaStream_t stream)
+{
+  // no-op
+}
+
+template <typename InputIterator, typename OutputIterator>
+std::enable_if_t<
+  std::is_arithmetic<typename std::iterator_traits<OutputIterator>::value_type>::value,
+  void>
+device_multicast_sendrecv_impl(raft::comms::comms_t const& comm,
+                               InputIterator input_first,
+                               std::vector<size_t> const& tx_counts,
+                               std::vector<size_t> const& tx_offsets,
+                               std::vector<int> const& tx_dst_ranks,
+                               OutputIterator output_first,
+                               std::vector<size_t> const& rx_counts,
+                               std::vector<size_t> const& rx_offsets,
+                               std::vector<int> const& rx_src_ranks,
+                               cudaStream_t stream)
+{
+  using value_type = typename std::iterator_traits<InputIterator>::value_type;
+  static_assert(
+    std::is_same<typename std::iterator_traits<OutputIterator>::value_type, value_type>::value);
+  // ncclSend/ncclRecv pair needs to be located inside ncclGroupStart/ncclGroupEnd to avoid deadlock
+  ncclGroupStart();
+  for (size_t i = 0; i < tx_counts.size(); ++i) {
```

Comment:
Consider replacing custom loops with transform/for_each algorithms (especially for loops with only a few calls inside).